src/main/java/org/apache/maven/plugins/shade/filter/MinijarFilter.java

```diff
@@ -156,6 +204,43 @@ private void removeServices( final MavenProject project, final Clazzpath cp )
         while ( repeatScan );
     }
 
+    private boolean removeServicesFromDir( Clazzpath cp, Set<Clazz> neededClasses, String fileName )
+    {
+        final File servicesDir = new File( fileName, "META-INF/services/" );
+        if ( !servicesDir.isDirectory() )
+        {
+            return false;
+        }
+        final File[] serviceProviderConfigFiles = servicesDir.listFiles();
+        if ( serviceProviderConfigFiles == null || serviceProviderConfigFiles.length == 0 )
+        {
+            return false;
+        }
+
+        boolean repeatScan = false;
+        for ( File serviceProviderConfigFile : serviceProviderConfigFiles )
+        {
+            final String serviceClassName = serviceProviderConfigFile.getName();
+            final boolean isNeededClass = neededClasses.contains( cp.getClazz( serviceClassName ) );
+            if ( !isNeededClass )
+            {
+                continue;
+            }
+
+            try ( final BufferedReader configFileReader = new BufferedReader(
+                    new InputStreamReader( new FileInputStream( serviceProviderConfigFile ), UTF_8 ) ) )
+            {
+                // check whether the found classes use services in turn
+                repeatScan |= scanServiceProviderConfigFile( cp, configFileReader );
+            }
+            catch ( final IOException e )
+            {
+                log.warn( e.getMessage() );
```

Comment:
I think this needs a decent message and in warn just the log the message and when in verbose mode also add the exception.

Reply:
I did it the same way @gnodet committed it a few lines down in `removeServicesFromJar`. If one gets a decent message, then the other should, too. What would you have me do?

Reply:
Agreed, make it consistent with the rest. Improving those is a different topic.

Reply:
Actually, it was easy enough to comply with your wish (see the new commit). Secondly, Guillaume's authorship of the code comes from his squashing one of my previous PRs, so actually it is fair enough for me to improve both places.